lib/EPub.php

```diff
@@ -0,0 +1,843 @@
+        if (!$coverid) {
+            return $this->no_cover();
+        }
+
+        $nodes = $this->xpath->query('//opf:manifest/opf:item[@id="' . $coverid . '"]');
+        if (!$nodes->length) {
+            return $this->no_cover();
+        }
+        $mime = $nodes->item(0)->attr('opf:media-type');
+        $path = $nodes->item(0)->attr('opf:href');
+        $path = dirname('/' . $this->meta) . '/' . $path; // image path is relative to meta file
+        $path = ltrim($path, '/');
+
+        $zip = new ZipArchive();
+        if (!@$zip->open($this->file)) {
+            throw new Exception('Failed to read epub file');
+        }
+        $data = $zip->getFromName($path);
+
+        return array(
+            'mime'  => $mime,
+            'data'  => $data,
+            'found' => $path
+        );
+    }
+
+    public function getCoverItem()
+    {
+        $nodes = $this->xpath->query('//opf:metadata/opf:meta[@name="cover"]');
+        if (!$nodes->length) {
+            return NULL;
+        }
+
+        $coverid = (String)$nodes->item(0)->attr('opf:content');
+        if (!$coverid) {
+            return NULL;
+        }
+
+        $nodes = $this->xpath->query('//opf:manifest/opf:item[@id="' . $coverid . '"]');
+        if (!$nodes->length) {
+            return NULL;
+        }
+
+        return $nodes->item(0);
+    }
+
+    public function Combine($a, $b)
+    {
+        $isAbsolute = false;
+        if ($a[0] == '/') {
+            $isAbsolute = true;
+        }
+
+        if ($b[0] == '/') {
+            throw new InvalidArgumentException('Second path part must not start with ' . $m_Separator);
+        }
+
+        $splittedA = preg_split('#/#', $a);
+        $splittedB = preg_split('#/#', $b);
+
+        $pathParts = array();
+        $mergedPath = array_merge($splittedA, $splittedB);
+
+        foreach ($mergedPath as $item) {
+            if ($item == null || $item == '' || $item == '.') {
+                continue;
+            }
+
+            if ($item == '..') {
+                array_pop($pathParts);
+                continue;
+            }
+
+            array_push($pathParts, $item);
+        }
+
+        $path = implode('/', $pathParts);
+        if ($isAbsolute) {
+            return('/' . $path);
+        } else {
+            return($path);
+        }
+    }
+
+    private function getFullPath($file, $context = NULL)
+    {
+        $path = dirname('/' . $this->meta) . '/' . $file;
+        $path = ltrim($path, '\\');
+        $path = ltrim($path, '/');
+        if (!empty($context)) {
+            $path = $this->combine(dirname($path), $context);
+        }
+        //error_log ("FullPath : $path ($file / $context)");
+        return $path;
+    }
+
+    public function updateForKepub()
+    {
+        $item = $this->getCoverItem();
+        if (!is_null($item)) {
+            $item->attr('opf:properties', 'cover-image');
+        }
+    }
+
+    public function Cover2($path=false, $mime=false)
```

Comment:
Cover2 started as a copy of Cover because my first need was only to replace an already existing cover by a new one. Since I never updated Cover2 to also create a cover in a Epub that didn't have one in the first place, I kept it here as a reminder.